File: tests/traits.py
# """
# Examples of classes that derive from other classes. These could
# use abstract base classes to enforce the implementation of methods
# in derived classes, but that is not essential to the pattern.
# """

class Waddle:
    def waddle(self, elegant: bool):
        # """
        # Waddle like a duck, either elegantly or not.
        # """
        pass

    def run(self, speed: int):
        # """
        # Ducks can also run
        # """
        pass

class Quack:
    def quack(self, volume: float):
        # """
        # Quack like a duck
        # """
        pass

    def echoes(self) -> bool:
        # """
        # Does this quack echo?
        # """
        pass

class Duck(Waddle, Quack):
    def __init__(self, echoes: bool):
        self._echoes = echoes

    def waddle(self, elegant: bool):
        if elegant:
            print("elegantly ", end='')
        print("waddle")

    def run(self, speed: int):
        print("run", speed)

    def quack(self, volume: float):
        print("quack", volume)

    def echoes(self) -> bool:
        return self._echoes

    def fly(self):
        print("fly")

def make_it_waddle(waddler: Waddle):
    waddler.waddle(False)

def make_it_quack(quacker: Quack):
    quacker.quack(10.0)

def tests():
    donald = Duck(True)
    make_it_waddle(donald)
    make_it_quack(donald)

File: src/traits.rs
trait Waddle {
    fn waddle(&self, elegant: bool);
    fn run(&self, speed: i64);
}

trait Quack {
    fn quack(&self, volume: f64);
    fn echoes(&self) -> bool;
}

pub struct Duck {
    pub _echoes: bool,
}

impl Duck {
    pub fn new(echoes: bool) -> Duck {
        let tmp__echoes = echoes;
        Duck {
            _echoes: tmp__echoes,
        }
    }

    pub fn fly(&self) {
        println!("fly");
    }

}

impl Waddle for Duck {
    fn waddle(&self, elegant: bool) {
        if elegant {
            print!("elegantly ");
        }
        println!("waddle");
    }

    fn run(&self, speed: i64) {
        println!("{} {}", "run", speed);
    }

}

impl Quack for Duck {
    fn quack(&self, volume: f64) {
        println!("{} {}", "quack", volume);
    }

    fn echoes(&self) -> bool {
        return self._echoes;
    }

}

pub fn make_it_waddle(waddler: &mut Waddle) {
    waddler.waddle(false);
}

pub fn make_it_quack(quacker: &mut Quack) {
    quacker.quack(10.0);
}

pub fn tests() {
    let donald = Duck::new(true);
    make_it_waddle(donald);
    make_it_quack(donald);
}

